usr/src/uts/common/io/scsi/adapters/smartpqi/smartpqi_init.c

```diff
@@ -0,0 +1,1948 @@
+			goto error_out;
+		new_list_len =
+		    ntohl(((report_lun_header_t *)data)->list_length);
+		new_data_len = sizeof (report_lun_header_t) +
+		    new_list_len;
+	} while (new_list_len > list_len);
+	rval = B_TRUE;
+
+error_out:
+	if (rval == B_FALSE) {
+		PQI_FREE(data, data_len);
+		data = NULL;
+	}
+	*buf = data;
+	return (rval);
+}
+
+static inline boolean_t
+report_phys_luns(pqi_state_t s, void **v)
+{
+	return (report_luns_by_cmd(s, CISS_REPORT_PHYS, v));
+}
+
+static inline boolean_t
+report_logical_luns(pqi_state_t s, void **v)
+{
+	return (report_luns_by_cmd(s, CISS_REPORT_LOG, v));
+}
+
+static boolean_t
+get_device_list(pqi_state_t s, report_phys_lun_extended_t **pl,
+    report_log_lun_extended_t **ll)
+{
+	report_log_lun_extended_t	*log_data;
+	report_log_lun_extended_t	*internal_log;
+	size_t				list_len;
+	size_t				data_len;
+	report_lun_header_t		header;
+
+	if (report_phys_luns(s, (void **)pl) == B_FALSE)
+		return (B_FALSE);
+
+	if (report_logical_luns(s, (void **)ll) == B_FALSE)
+		return (B_FALSE);
+
+	log_data = *ll;
+	if (log_data) {
+		list_len = ntohl(log_data->header.list_length);
+	} else {
+		(void) memset(&header, 0, sizeof (header));
+		log_data = (report_log_lun_extended_t *)&header;
+		list_len = 0;
+	}
+
+	data_len = sizeof (header) + list_len;
+	/*
+	 * Add the controller to the logical luns which is a empty device
+	 */
+	internal_log = PQI_ZALLOC(data_len +
+	    sizeof (report_log_lun_extended_entry_t), KM_SLEEP);
+	(void) memcpy(internal_log, log_data, data_len);
+	internal_log->header.list_length = htonl(list_len +
+	    sizeof (report_log_lun_extended_entry_t));
+
+	if (*ll != NULL)
+		PQI_FREE(*ll, sizeof (report_lun_header_t) +
+		    ntohl((*ll)->header.list_length));
+	*ll = internal_log;
+	return (B_TRUE);
+}
+
+/* ---- Only skip physical devices ---- */
+static boolean_t
+skip_device(char *addr)
+{
+	return (MASKED_DEVICE(addr) ? B_TRUE : B_FALSE);
+}
+
+static boolean_t
+get_device_info(pqi_state_t s, pqi_device_t dev)
+{
+	boolean_t		rval = B_FALSE;
+	struct scsi_inquiry	*inq;
+
+	inq = PQI_ZALLOC(sizeof (*inq), KM_SLEEP);
+	if (pqi_scsi_inquiry(s, dev, 0, inq, sizeof (*inq)) == B_FALSE)
+		goto out;
+
+	dev->pd_devtype = inq->inq_dtype & 0x1f;
+	(void) memcpy(dev->pd_vendor, inq->inq_vid, sizeof (dev->pd_vendor));
+	(void) memcpy(dev->pd_model, inq->inq_pid, sizeof (dev->pd_model));
+
+	/* TODO Handle logical devices */
+	rval = B_TRUE;
+out:
+	PQI_FREE(inq, sizeof (*inq));
+	return (rval);
+}
+
+static boolean_t
+is_supported_dev(pqi_state_t s, pqi_device_t dev)
+{
+	boolean_t	rval = B_FALSE;
+
+	switch (dev->pd_devtype) {
+	case DTYPE_DIRECT:
+	case TYPE_ZBC:
+	case DTYPE_SEQUENTIAL:
+	case DTYPE_ESI:
+		rval = B_TRUE;
+		break;
+	case DTYPE_ARRAY_CTRL:
+		if (strncmp(dev->pd_scsi3addr, RAID_CTLR_LUNID,
+		    sizeof (dev->pd_scsi3addr)) == 0)
+			rval = B_TRUE;
+		break;
+	default:
+		dev_err(s->s_dip, CE_WARN, "Not supported device: 0x%x\n",
```

Comment:
Would be better with the device type stringified, so the user could more readily see _what_ was not a supported device.

Also needs to be "Not a supported ..." or "%s is not a supported..." or something grammatical.